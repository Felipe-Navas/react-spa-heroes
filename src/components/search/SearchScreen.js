import { useMemo } from 'react';
import queryString from 'query-string';
import { useNavigate, useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../hero/HeroCard';

export const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = '' } = queryString.parse(location.search);

  const [formValues, handleInputChange] = useForm({ searchText: q });

  const { searchText } = formValues;

  const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${searchText}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Search form</h4>
          <hr />
          <form onSubmit={handleSearch}>
            <input
              type="text"
              placeholder="Find a hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={handleInputChange}
            />
            <button type="submit" className="btn btn-outline-primary mt-1">
              Find
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Resultados</h4>
          <hr />

          {q === '' ? (
            <div className="alert alert-info">Find a Hero </div>
          ) : (
            heroesFiltered.length === 0 && (
              <div className="alert alert-danger">No results: {q}</div>
            )
          )}

          {heroesFiltered.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </>
  );
};
