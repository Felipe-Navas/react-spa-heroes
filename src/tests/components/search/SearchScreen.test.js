import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('Testing the <SearchScreen /> component', () => {
  test('should render correctly with default values', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Find a Hero');
  });

  test('should show Batman and the input with the value of the queryString', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').prop('value')).toBe('batman');
    expect(wrapper.find('.card').length).toBe(1);
    expect(wrapper.find('.card-title').text().trim()).toBe('Batman');
  });

  test('should show the error message when the hero is not found', () => {
    const query = 'notfound';
    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${query}`]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper.find('input').prop('value')).toBe('notfound');
    expect(wrapper.find('.alert-danger').text().trim()).toBe(
      `No results: ${query}`
    );
  });

  test('should call the navigate method and show the new component', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );

    const input = wrapper.find('input');
    input.simulate('change', {
      target: { name: 'searchText', value: 'batman' },
    });

    const form = wrapper.find('form');
    // * This is a valid way too, to call the submit method
    // form.simulate('submit', { preventDefault: () => {} });

    form.prop('onSubmit')({ preventDefault: () => {} });

    expect(mockNavigate).toHaveBeenCalled();
    expect(mockNavigate).toHaveBeenCalledWith('?q=batman');
  });
});
