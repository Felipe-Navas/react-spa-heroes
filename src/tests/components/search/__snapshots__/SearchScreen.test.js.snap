// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the <SearchScreen /> component should render correctly with default values 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search a hero by name
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find a hero"
          type="text"
          value=""
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-outline-primary mt-1"
            type="submit"
          >
            Find
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Find a Hero 
      </div>
    </div>
  </div>,
]
`;

exports[`Testing the <SearchScreen /> component should show Batman and the input with the value of the queryString 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search a hero by name
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find a hero"
          type="text"
          value="batman"
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-outline-primary mt-1"
            type="submit"
          >
            Find
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
